src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,571 @@
+      const device = new LocalDevice(rest);
+      device.loadPersisted();
+      return device;
+    }
+
+    loadPersisted() {
+      const Platform = this.rest.Platform;
+      if (!Platform.Config.push) {
+        throw new this.rest.ErrorInfo('Push activation is not available on this platform', 40000, 400);
+      }
+      this.platform = Platform.Config.push.platform;
+      this.clientId = this.rest.auth.clientId ?? undefined;
+      this.formFactor = Platform.Config.push.formFactor;
+      this.id = Platform.Config.push.storage.get(persistKeys.deviceId);
+
+      if (this.id) {
+        this.deviceSecret = Platform.Config.push.storage.get(persistKeys.deviceSecret) || undefined;
+        this.deviceIdentityToken = JSON.parse(
+          Platform.Config.push.storage.get(persistKeys.deviceIdentityToken) || 'null',
+        );
+        this.push.recipient = JSON.parse(Platform.Config.push.storage.get(persistKeys.pushRecipient) || 'null');
+      } else {
+        this.resetId();
+      }
+    }
+
+    persist() {
+      const config = this.rest.Platform.Config;
+      if (!config.push) {
+        throw new this.rest.ErrorInfo('hello', 1, 1);
+      }
+      if (this.id) {
+        config.push.storage.set(persistKeys.deviceId, this.id);
+      }
+      if (this.deviceSecret) {
+        config.push.storage.set(persistKeys.deviceSecret, this.deviceSecret);
+      }
+      if (this.deviceIdentityToken) {
+        config.push.storage.set(persistKeys.deviceIdentityToken, JSON.stringify(this.deviceIdentityToken));
+      }
+      if (this.push.recipient) {
+        config.push.storage.set(persistKeys.pushRecipient, JSON.stringify(this.push.recipient));
+      }
+    }
+
+    resetId() {
+      this.id = ulid();
+      this.deviceSecret = ulid();
+      this.persist();
+    }
+
+    getAuthDetails(
+      rest: BaseRest,
+      headers: Record<string, string>,
+      params: Record<string, unknown>,
+    ): LocalDeviceAuthDetails {
+      if (!this.deviceIdentityToken) {
+        throw new this.rest.ErrorInfo('Unable to update device registration; no deviceIdentityToken', 50000, 500);
+      }
+      if (this.rest.http.supportsAuthHeaders) {
+        return {
+          headers: rest.Utils.mixin(
+            { authorization: 'Bearer ' + rest.Utils.toBase64(this.deviceIdentityToken) },
+            headers,
+          ) as Record<string, string>,
+          params,
+        };
+      } else {
+        return { headers, params: rest.Utils.mixin({ access_token: this.deviceIdentityToken }, params) };
+      }
+    }
+  };
+}
+
+export class ActivationStateMachine {
+  rest: BaseRest;
+  current: ActivationState;
+  pendingEvents: ActivationEvent[];
+  handling: boolean;
+  deactivatedCallback?: ErrCallback;
+  activatedCallback?: ErrCallback;
+  pushConfig: IPlatformPushConfig;
+  registerCallback?: RegisterCallback;
+  deregisterCallback?: DeregisterCallback;
+  updatedCallback?: ErrCallback;
+
+  // Used for testing
+  pushManager?: PushManager;
```

Comment:
maybe it's worth to add `_` at the begging just to make it clear that it's not a part of public API